### ISSUE TITLE
Feature request: shared (essential) plugins

### ISSUE THREAD

**Issue by AvianY** [vim-volt/volt#208]
I am interested what do you think of the idea of volt defining a special set of plugins which are loaded regardless of which profile you are in.
I noticed that there are many so essential plugins that I use, whom I need always to be active. Such plugins could simply be put in the ../pack/*/start/ directory. Volt could also update them when requested. In the end, only the plugins which are different would be loaded into vim.
Maybe an even better idea would be for volt to define "plugin sets", where for each profile you would choose 2 or more sets which are then loaded. One of these sets could be the "default" set, if the user adds it to this list of sets... just an idea.
as it is right now, I would have to know the subset of my plugins which are essential if I wanted to always copy them to the new profile I create. Also if I decide to add/remove an essential plugin, I have to do the same across all of the profiles.
Could something like this be possible to implement? I think it would greatly improve on the idea of profiles in volt.

**Comment by AvianY** [comment]
Are you saying that this already works, or that this will be implemented?
I can't get the last command to work...

**Comment by tyru** [comment]
Currently volt saves repository's paths as `{domain}/{user}/{plugin}` format in lock.json .
But that should be saved as URIs, maybe.
Supported protocols are only `https`, `http`, `git`, `ssh`, `profile`.
(btw, `ssh` may not work now, I don't test it yet though)

**Comment by tyru** [comment]
> Are you saying that this already works, or that this will be implemented?

Ah sorry, it is just an idea.
Not implemented yet.

**Comment by AvianY** [comment]
Yes, I think this idea works great for sharing common packages in case you are adding them.
But let's say that I am removing a package from the 'basic' profile.... How could I apply the same change across all other profiles, for which the 'basic' profile is a subset of them.

**Comment by tyru** [comment]
I'm thinking data model like this in lock.json:

```json
  "profiles": [
    {
      "name": "default",
      "repos_path": [
        "profile://basic",
        "https://github.com/tyru/caw.vim"
      ]
    },
    {
      "name": "basic",
      "repos_path": [
        "https://github.com/tpope/vim-sensible",
        "https://github.com/kana/vim-textobj-user",
        "https://github.com/kana/vim-textobj-indent"
      ]
    }
  ]
```

default profile can use "basic profile + tyru/caw.vim".
and removing vim-sensible by `volt profile rm basic tpope/vim-sensible` (or `volt rm tpope/vim-sensible` if current profile is basic) makes `tpope/vim-sensible` unavailable in default profile too.
And also other profiles which reference basic profile.

**Comment by AvianY** [comment]
Yes, great idea! so basically, a profile can equally contain plugins and other profiles, so you can nest them. This is neat! :+1:

**Comment by tyru** [comment]
Volt must disallow recursive reference (below is directed graphs).

* A <---> A
* A <---> B
* A ---> B, B ---> C, C ---> A

**Comment by tyru** [comment]
> Volt must disallow recursive reference (below is directed graphs).
> 
> * A <---> A
> * A <---> B
> * A ---> B, B ---> C, C ---> A

## Algorithm

Just recursively find the same profile name at the start.

* A <---> A
  * A --->Found A
* A <---> B
  * A ---> B --> Found A
* A ---> B, B ---> C, C ---> A
  * A ---> B ---> C ---> Found A

This is naive algorithm.
This does not scale but is simple.
More efficient algorithm may be needed in the future.

**Comment by AvianY** [comment]
I was wondering what would the command `$volt list` show ... ?
Would it be ok, that by itself it shows the top level plugins and other profiles that are included?
If the user desires to view top level plugins as well as all the plugins added with other profiles, an additional switch `-e` (expand) could be introduced. Like so `$volt list -e`.

what do you think?

**Comment by tyru** [comment]
The question seems unrelated to this issue.
Please file a new issue.
